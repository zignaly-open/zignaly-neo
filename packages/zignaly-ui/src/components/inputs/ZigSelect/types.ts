import { MenuPlacement, Props as SelectProps } from "react-select";

export type ZigSelectOption<T> = {
  value: T;
  label: string | JSX.Element;
  disabled?: boolean;
};

export type ZigSelectProps<T> = Omit<
  SelectProps,
  "isDisabled" | "isOptionDisabled" | "onChange" | "options"
> & {
  label?: string | JSX.Element;
  id?: string;
  small?: boolean;
  outlined?: boolean;
  options?: ZigSelectOption<T>[];
  filterOption?: ({ data }: { data: ZigSelectOption<T> }, search: string) => boolean;
  onChange?: (value: ZigSelectOption<T>["value"] | null, option: ZigSelectOption<T> | null) => void;
  value?: ZigSelectOption<T>["value"];
  width?: number;
  error?: string;
  placeholder?: string;
  disabled?: boolean;
  menuPlacement?: MenuPlacement;
  helperText?: string;
};
