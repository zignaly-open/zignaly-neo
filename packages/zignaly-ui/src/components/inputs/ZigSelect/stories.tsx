import React, { useState } from "react";
import "@mui/system";
import { ComponentMeta, ComponentStory } from "@storybook/react";
import ZigSelect from "./index";
import "@mui/system";

const result: ComponentMeta<typeof ZigSelect> = {
  title: "Inputs/ZigSelect",
  component: ZigSelect,
  argTypes: {
    label: {
      type: "string",
    },
    placeholder: {
      type: "string",
    },
    disabled: {
      type: "boolean",
      defaultValue: false,
    },
    small: {
      type: "boolean",
      defaultValue: false,
    },
    defaultValue: {
      type: "string",
    },
    error: {
      type: "string",
      defaultValue: "",
    },
  },
};

export default result;

const options = [
  { label: "Regular cat", value: 1 },
  { label: "Regular dog", value: -1 },
  {
    label: (
      <>
        <img
          height={30}
          width={30}
          src={
            "https://cdna.artstation.com/p/assets/images/images/014/532/698/20181209121125/smaller_square/catgirl-enthusiast-girl-4.jpg?1544379086"
          }
        />{" "}
        Catgirl
      </>
    ),
    value: 2,
  },
];

const Template: ComponentStory<typeof ZigSelect> = (args) => {
  const [value, setValue] = useState<number>();

  return (
    <>
      Value: {JSON.stringify(value)}
      {value === 2 ? " Meow" : ""}
      <br />
      <ZigSelect
        disabled={args.disabled}
        placeholder={args.placeholder}
        error={args.error}
        label={args.label}
        value={value}
        small={args.small}
        onChange={(v) => setValue(v!)}
        options={options}
      />
    </>
  );
};

export const BasicSelect: ComponentMeta<typeof ZigSelect> = Template.bind({});
BasicSelect.args = {
  label: "Choose pet",
  placeholder: "Choose pet",
};
