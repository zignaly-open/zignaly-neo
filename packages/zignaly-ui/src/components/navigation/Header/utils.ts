// TODO: Move this for config or theme folder.
export const headerConfig = {
  height: "52px",
};
