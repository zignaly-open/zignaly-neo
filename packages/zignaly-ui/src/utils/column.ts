import styled from "styled-components";

export const Column = styled.div<{
  gap?: number;
  justifyContent?: "center" | "end" | "start" | "space-between";
  alignItems?: "center" | "end" | "start" | "space-between";
  textAlign?: "end" | "center" | "start";
  height?: string;
  flex?: number;
  width?: string;
}>`
  display: flex;
  flex-direction: column;
  flex: ${(props) => props.flex};
  gap: ${(props) => props.gap}px;
  justify-content: ${(props) => props.justifyContent};
  align-items: ${(props) => props.alignItems};
  text-align: ${(props) => `${props.textAlign}`};
  height: ${(props) => `${props.height}`};
  width: ${(props) => `${props.width}`};
`;
