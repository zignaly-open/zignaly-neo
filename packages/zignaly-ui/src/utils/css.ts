import { css } from "styled-components";

export const NiceScrollbar = css`
  &::-webkit-scrollbar {
    width: 6px;
  }

  /* Track */
  &::-webkit-scrollbar-track {
    background: rgba(255, 255, 255, 0.05);
    border-radius: 12px;
  }

  /* Handle */
  &::-webkit-scrollbar-thumb {
    background: rgba(255, 255, 255, 0.1);
    border-radius: 12px;
  }

  /* Handle on hover */
  &::-webkit-scrollbar-thumb:hover {
    background: rgba(255, 255, 255, 0.15);
  }
`;
