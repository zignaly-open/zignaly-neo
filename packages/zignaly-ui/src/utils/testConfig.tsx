import { render } from "@testing-library/react";
import { ThemeProvider as ThemeProviderMui } from "@mui/material";
import { ThemeProvider, ThemeProviderMui as ThemeInheritorMui } from "../index";
import React from "react";
import { getZignalyThemeExport } from "../theme";
import mediaQuery from "css-mediaquery";
import { JestI18nextDecorator } from "./dummyI18nextProvider";

export const renderWithProvidersUi = (ui: JSX.Element, renderOptions = {}) => {
  const { mui, legacyStyledComponentsDoNotUse } = getZignalyThemeExport();
  return render(ui, {
    wrapper: ({ children }: { children: JSX.Element }) => (
      <JestI18nextDecorator>
        <ThemeProvider theme={legacyStyledComponentsDoNotUse}>
          <ThemeInheritorMui theme={mui}>
            <ThemeProviderMui theme={mui}>{children}</ThemeProviderMui>
          </ThemeInheritorMui>
        </ThemeProvider>
      </JestI18nextDecorator>
    ),
    ...renderOptions,
  });
};

const createMatchMedia = (width: number) => {
  return (query: string) => {
    return {
      matches: mediaQuery.match(query, { width }),
      media: "",
      addListener: () => {},
      removeListener: () => {},
      onchange: () => {},
      addEventListener: () => {},
      removeEventListener: () => {},
      dispatchEvent: () => true,
    };
  };
};

export const resizeScreenSize = (width: number) => {
  window.matchMedia = createMatchMedia(width);
};
