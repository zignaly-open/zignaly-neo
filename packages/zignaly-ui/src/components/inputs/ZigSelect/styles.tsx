import { Box, GlobalStyles, styled } from "@mui/material";
import { NiceScrollbar } from "utils/css";
import React from "react";
import { css } from "@emotion/react";

type Props = { error?: string; width?: number; small?: boolean; outlined?: boolean };

export const StyledSelectWrapper = styled(Box)<Props>`
  .zig-react-select {
    &__control {
      border: 1px solid
        ${({ theme, error }) => (error ? theme.palette.redGraphOrError : theme.palette.neutral600)};
      padding: ${({ small }) => (small ? "3px 16px 3px 9px" : "11px 24px 11px 16px")};
      margin-top: 4px;
      min-height: ${({ small }) => (small ? "30.75px" : "60px")};
      border-radius: 5px;
      display: flex;
      align-items: center;
      ${({ outlined }) =>
        css`
          background: ${outlined ? "transparent" : "rgba(16, 18, 37)"};
          background: ${outlined
            ? "transparent"
            : `linear-gradient(
            90deg,
            rgb(16 18 37) 0%,
            rgb(16 18 37) 35%,
            rgb(16 18 37) 100%
          );`};
        `}
      transition: border-color 0.2s;
      margin-bottom: 3px;

      ${({ width }) => width && `width: ${width}${width?.toString().includes("%") ? "" : "px"}`};

      padding-right: 0;

      &:hover {
        border-color: ${({ theme, error }) =>
          error ? theme.palette?.redGraphOrError : theme.palette.neutral400};
        ${({ outlined }) =>
          outlined &&
          css`
            background-color: rgba(118, 130, 247, 0.08);
          `}
        ${({ outlined, theme }) =>
          outlined &&
          css`
            .zig-react-select__single-value {
              color: ${theme.palette.neutral000} !important;
            }
          `}
      }

      &--is-focused {
        border: 1px solid
          ${({ theme, error }) =>
            error ? theme.palette.redGraphOrError : theme.palette.neutral400};
        box-shadow: none !important;
      }
    }

    &__placeholder,
    &__single-value {
      padding: 0;
      margin-left: 0;
    }

    &__value-container {
      ${({ small }) =>
        small &&
        css`
          padding: 0 5px;
        `}
    }

    &__placeholder,
    &__single-value,
    &__input-container,
    &__input {
      font-size: ${({ small }) => (small ? "13px" : "16px")};
      line-height: ${({ small }) => (small ? "15px" : "20px")};
    }

    &__single-value,
    &__input {
      color: ${({ theme }) => theme.palette.neutral100} !important;
    }

    .zig-react-select__menu {
      border: 1px solid ${({ theme }) => theme.palette.neutral600} !important;
      color: ${({ theme }) => theme.palette.neutral200} !important;
      background: rgba(16, 18, 37) !important;
    }

    &__placeholder {
      color: ${({ theme }) => theme.palette.neutral400} !important;
    }

    &__input-container {
      margin-left: 0;
    }

    &__indicator {
      ${({ width }) =>
        width &&
        width <= 100 &&
        css`
          padding: 0 2px;
        `};
      ${({ small }) =>
        small &&
        css`
          padding: 0 4px;
          width: 22px;
          height: 22px;
        `};
    }
  }
` as React.FC<Props>;

export const ZigSelectGlobalStyle = (
  <GlobalStyles
    styles={css`
      .zig-react-select {
        &__menu {
          &-list {
            ${NiceScrollbar.toString()};
          }
        }
      }
    `}
  />
);
