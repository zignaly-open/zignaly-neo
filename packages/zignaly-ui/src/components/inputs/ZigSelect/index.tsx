import React from "react";
import { StyledSelectWrapper, ZigSelectGlobalStyle } from "./styles";
import { ZigSelectOption, ZigSelectProps } from "./types";
import ErrorMessage from "../../display/ErrorMessage";
import { Typography } from "@mui/material";
import Select from "react-select";

function ZigSelect<T>({
  onChange,
  value,
  label,
  error,
  helperText,
  width,
  placeholder,
  options,
  small,
  disabled,
  ...props
}: ZigSelectProps<T>): JSX.Element {
  return (
    <StyledSelectWrapper error={error} width={width} small={small}>
      {label && <Typography color={"neutral200"}>{label}</Typography>}
      {ZigSelectGlobalStyle}
      <Select
        components={{
          IndicatorSeparator: () => null,
        }}
        // if you want to use this inside of a modal, pass it `menuPosition="fixed"`, `menuShouldScrollIntoView={false}` and `menuShouldBlockScroll`
        isOptionDisabled={(option) => !!(option as ZigSelectOption<T>).disabled}
        options={options as unknown as { label: string; value: number }[]}
        isDisabled={disabled}
        onChange={(v) => {
          onChange?.((v as ZigSelectOption<T>)?.value ?? null, (v as ZigSelectOption<T>) || null);
        }}
        menuPortalTarget={document.body}
        placeholder={placeholder || label}
        value={options?.find?.((x) => x.value === value || (x as unknown) === value) || null}
        classNamePrefix="zig-react-select"
        styles={{
          singleValue: (provided, state) => ({
            ...provided,
            display: state.selectProps.menuIsOpen ? "none" : "block",
          }),
        }}
        {...props}
      />
      {!!error && <ErrorMessage text={error} />}
    </StyledSelectWrapper>
  );
}

export default ZigSelect;
