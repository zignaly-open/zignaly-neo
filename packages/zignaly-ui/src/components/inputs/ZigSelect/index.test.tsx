import React from "react";
import { renderWithProvidersUi } from "../../../utils/testConfig";
import { fireEvent, waitFor, screen } from "@testing-library/react";
import ZigSelect from ".";

describe("components/inputs/ZigSelect", () => {
  const options = [
    { label: "Option 1", value: 1 },
    { label: "Option 2", value: 2 },
    { label: "Option 3", value: 3 },
  ];

  it("should render without crashing", () => {
    const { container } = renderWithProvidersUi(<ZigSelect options={options} id={"test-id"} />);

    const select = container.querySelector("#test-id") as HTMLElement;
    expect(select).toBeVisible();
    expect(select).toMatchSnapshot();
  });

  it("should open dropdown on click", async () => {
    const { container } = renderWithProvidersUi(<ZigSelect options={options} id={"test-id"} />);

    const dropdownIndicator = container.querySelector("#react-select-test-id-input") as HTMLElement;
    fireEvent.mouseDown(dropdownIndicator);

    await waitFor(() => {
      options.map((optionText) => {
        const option = screen.getByText(optionText.label);
        expect(option).toBeVisible();
      });
    });
  });

  it("should trigger onChange for the selected option", async () => {
    const onChangeMock = jest.fn();
    const { container } = renderWithProvidersUi(
      <ZigSelect options={options} onChange={onChangeMock} id={"test-id"} />,
    );

    const dropdownIndicator = container.querySelector("#react-select-test-id-input") as HTMLElement;
    fireEvent.mouseDown(dropdownIndicator);

    await waitFor(() => {
      const option1 = screen.getByText("Option 1");
      fireEvent.click(option1);
    });

    await waitFor(() => {
      expect(onChangeMock).toHaveBeenCalledWith(1, expect.anything());
    });
  });

  it("should display error message", () => {
    renderWithProvidersUi(<ZigSelect options={options} error="Invalid option" />);

    const errorMessage = screen.getByText("Invalid option");
    expect(errorMessage).toBeVisible();
  });
});
