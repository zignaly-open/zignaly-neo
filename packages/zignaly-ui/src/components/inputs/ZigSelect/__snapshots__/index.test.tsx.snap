// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/inputs/ZigSelect should render without crashing 1`] = `
<div
  class=" css-b62m3t-container"
  id="test-id"
>
  <span
    class="css-1f43avz-a11yText-A11yText"
    id="react-select-test-id-live-region"
  />
  <span
    aria-atomic="false"
    aria-live="polite"
    aria-relevant="additions text"
    class="css-1f43avz-a11yText-A11yText"
  />
  <div
    class="zig-react-select__control css-1s2u09g-control"
  >
    <div
      class="zig-react-select__value-container css-dlobye-Component"
    >
      <div
        class="zig-react-select__placeholder css-14el2xx-placeholder"
        id="react-select-test-id-placeholder"
      >
        Select...
      </div>
      <div
        class="zig-react-select__input-container css-1u4kn00-Component"
        data-value=""
      >
        <input
          aria-autocomplete="list"
          aria-describedby="react-select-test-id-placeholder"
          aria-expanded="false"
          aria-haspopup="true"
          autocapitalize="none"
          autocomplete="off"
          autocorrect="off"
          class="zig-react-select__input"
          id="react-select-test-id-input"
          role="combobox"
          spellcheck="false"
          style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
          tabindex="0"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="zig-react-select__indicators css-ny0e4k-Component"
    >
      <div
        aria-hidden="true"
        class="zig-react-select__indicator zig-react-select__dropdown-indicator css-tlfecz-indicatorContainer"
      >
        <svg
          aria-hidden="true"
          class="css-tj5bde-Svg"
          focusable="false"
          height="20"
          viewBox="0 0 20 20"
          width="20"
        >
          <path
            d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
